093.mediawiki

```diff
@@ -0,0 +1,329 @@
+
+===Intended Protocol Flow===
+
+
+In the Bitcoin Cash Network, blocks are relayed from a peer possessing the block (the sender) to a peer requesting it (the receiver). The core idea behind Graphene is for the sender to avoid sending information that is already held by the receiver. Blocks are comprised of a header that is previously unknown to the receiver, and a collection of transactions that are likely to be already within the receiver's mempool. Therefore, to save bandwidth, Graphene attempts to avoid resending the actual transactions and transaction IDs. Instead, Graphene makes use of two ''set reconciliation'' data structures for expressing the list of transactions: [https://en.wikipedia.org/wiki/Bloom_filter Bloom filters] [[093.mediawiki#ref4|<nowiki>[4]</nowiki>]] and Invertible Bloom Lookup Tables (IBLTs) [[093.mediawiki#ref2|<nowiki>[2</nowiki>]][[093.mediawiki#ref3|<nowiki>,3]</nowiki>]]. The combination of the two data structures is more efficient than using either alone to express the transaction list, and it is typically more efficient than constructing a list from shortened transaction IDs.
+
+<A id='fig1'>
+{| class="wikitable" style="float:left"
+|[[File:media/buip093-messaging.png]] 
+|  A diagram of Graphene's messaging between the sender and receiver. ''M'' is the set of transactions that are in the block but missing from the receiver mempool. If ''M'' is too large, then the IBLT will not decode; in this case a fail-over block must be requested.
+|}</a>
+
+An overview of Graphene is presented in [[093.mediawiki#fig1|Figure 1]], which depicts the following steps between two peers that have confirmed that they are running the same version of the protocol. 
+
+# A sender informs the receiver that a new block is available using an '''inv''' message. 
+# If the block is not already known to the receiver, she responds with a '''get_grblk''' message that includes a count of transactions in her mempool, ''m''. (Note: the count should also include the number of transactions in the receiver's orphan pool, but for brevity we will henceforth assume that the mempool includes transactions in the orphan pool as well).
+# The sender creates a Bloom filter ''S'' of all transaction IDs in the block along with an IBLT ''I'' containing just the "cheap hash" (first 8 bytes of the hash interpreted as an unsigned little endian 64 bit integer) of each transaction ID. Additionally, any full transactions likely missing from the receiver (such as the coinbase for the block) are collected into additional transactions list ''V''. If there does not exist a protocol-defined canonical transaction ordering, then the sender also creates a list of transaction ranks ''R'', listed lexicographically by ID, which indicates (with ''log(n)'' bits for each ID) the intended order for each transaction ID. The sender assembles ''S'', ''I'', ''V'', ''R'', and the block header into a '''grblk''' message, which he sends to the receiver.   
```

Comment:
How was 8-bytes chosen for the cheap hash size?  I think it makes sense to reduce Graphene's message size by using smaller hashes than 32-bytes, but if using cheap hashes is agreed to be the proper solution, then I think reducing to its minimal size is prudent.  1-byte is likely unreasonable, as the number of collisions for "small" blocks in the GB range is in the tens of thousands. However, 4-bytes may still be viable and further reduce the message size.

Reply:
It was chosen for abundance of caution with regard to the cheap hash collision attack. However, I can agree that smaller cheap hashes are probably pretty safe (especially considering the extra safety that comes from the Bloom filter as I described in my response to @deadalnix above). So I'm open to discussing a smaller fixed size or variable size as you suggested.